File: leetcode/rust-leetcode/src/2045.second-minimum-time-to-reach-destination.rs
use std::collections::VecDeque;
impl Solution {
    pub fn second_minimum(n: i32, edges: Vec<Vec<i32>>, time: i32, change: i32) -> i32 {
        let n = n as usize;
        let mut g = vec![vec![]; n+1];
        let (mut u, mut v, mut d) = (0, 0, 0);
        for e in edges {
            (u, v) = (e[0] as usize, e[1] as usize);
            g[u].push(v);
            g[v].push(u);
        }
        let mut q = VecDeque::new();
        // let mut least_time = 0;
        q.push_back((1, 0));
        let mut next_t = 0;
        let mut times = vec![[i32::MAX , i32::MAX]; n+1];
        times[1][0] = 0;
        while let Some((u, t)) = q.pop_front() {
            // if u == n {
            //     if least_time == 0 {
            //         least_time = t;
            //     } else if least_time < t {
            //         return t;
            //     }
            // }
            next_t = t;
            d = t / change;
            if d & 1 == 1 {
                next_t = (d + 1) * change;
            }
            next_t += time;
            for &v in g[u].iter() {
                if next_t < times[v][0] {
                    times[v][0] = next_t;
                    q.push_back((v, next_t));
                } else if times[v][0] < next_t && next_t < times[v][1] {
                    if v == n {
                        return next_t;
                    }
                    times[v][1] = next_t;
                    q.push_back((v, next_t));
                }
            }
        }
        0
    }
}

File: leetcode/go/2045.second-minimum-time-to-reach-destination.go
func secondMinimum(n int, edges [][]int, time int, change int) int {
    graph := make([][]int, n+1)
    for _, e := range edges {
        graph[e[0]] = append(graph[e[0]], e[1])
        graph[e[1]] = append(graph[e[1]], e[0])
    }
    times := make([][2]int, n+1)
    for i := range n+1 {
        times[i][0], times[i][1] = math.MaxInt32, math.MaxInt32
    }
    q := make([][2]int, 0, n)
    q = append(q, [2]int{1, 0})
    node := [2]int{0, 0}
    u, t, d, new_t := 0, 0, 0, 0
    for len(q) > 0 {
        node = q[0]
        q = q[1:]
        u, t = node[0], node[1]
        d = t/change
        if d & 1 == 1 {
            t = (d + 1) * change
        }
        new_t = t + time
        for _, v := range graph[u] {
            if new_t < times[v][0] {
                times[v][0] = new_t
                q = append(q, [2]int{v, new_t})
            } else if times[v][0] < new_t && new_t < times[v][1] {
                if v == n {
                    return new_t
                }
                times[v][1] = new_t
                q = append(q, [2]int{v, new_t})
            }
        }
    }
    return 0
}

